File: 0525-contiguous-array/0525-contiguous-array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        record = {}
        s = 0
        res = 0
        
        record[0] = -1
        for i in range(len(nums)):
            if nums[i]==1:
                curr = 1
            else:
                curr = -1
                
            s += curr
            
            if s not in record:
                record[s] = i
            else:
                res = max(res, i-record[s])
        
        return res

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int,int> M;
        int maxres = 0, cum = 0;
        M[0] = -1;
        for(int i=0;i<nums.size();i++){
            int curr = (nums[i]==1)?1:-1;
            cum += curr;
            if(!M.count(cum))
                M[cum] = i;
            else
                maxres = max(maxres,i-M[cum]);
        }
        return maxres;
    }
};